### PR TITLE
Feature/392 - Use AssetRenditions in ASC

### PR DESCRIPTION
**Pull request by davidjgonzalez**
This PR introduces the use of AssetRenditions (from previous PRs) to ASC. This includes:

1. Search results Card and List thumbnails
2. Cart list thumbnails
3. Details - Image component
4. Details - Video component

Note this does NOT include **Details - Renditions component** at this time, to retain some continuity between those Renditions and the Download renditions. 

In order to support the Details components, the following has been added:

1. An AssetRenditionsDataSource - which lists the AssetRenditions or selection by Authors, including the ability to filter which AssetRenditions can be displayed per data source. This datasource can be configured to:
   * Exclude renditions by name via `excludeAssetRenditions` property
   * Exclude all renditions provided by a named AssetRenditionDispatcher via `excludeAssetRenditionDispatchers` property
   * Include all renditions in AssetRenditionDispatchers that provide a matching "type" via `allowedAssetRenditionTypes` property. Leaving the type blank on the datasource OR the AssetRenditionDispatcher config will not restrict the listing (ie. if either type is blank, this will not result in filtering any renditions out of the datasource)    

2. A "legacy" dialog JavaScript, that allows the toggling of fieldsets in components, supporting a backward-compatible transition of components to the new paradigm. Note that NEW instances of the affected components will NOT allow them to be configured using legacy parameters; The legacy option ONLY is displayed if the component was previously configured with legacy configurations. This is primarily used internally in the ASC project itself, as customer implementations would/should not use both simultaneously in components they own.

### Search results and Card thumbs

This was a straightforward update and uses the Rendition ComputedProperty to select the `card` or `list` AssetRendition.  

```
<img src="${asset.properties['rendition?name=card']" .... />
```

Resulting in a URL of ... 

```
http://localhost:4502/content/dam/asset-share-commons/en/public/pictures/ruslan-bardash-351288.jpg.renditions/card/asset.rendition
```

#### No legacy config present

image

##### Dialog set to configure (non-Legacy mode) modern Asset Renditions

image

##### Dialog set to use Legacy mode

image

### Dispatcher considerations

Echoing Dispatcher consideration from AssetRendition PR, since the asset rendition URL "ends with" `.../asset.rendition`, ModMimeTypeInfo cannot be used to derive the response content type, and instead, the response headers should be cached by AEM Dispatcher.

### CODE REVIEW

File: core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/AssetRenditionImpl.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        if (computedPropertyParameters.get("download", false)) {
+            url += "download/";
+        }
+
+        url += "asset.rendition";
+
+        return escapeString(url);
+    }
+
+    private String escapeString(String str) {
```
Comment (bstopp) [inline core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/AssetRenditionImpl.java]: Isn't there more to escaping a string besides these?

File: core/src/main/java/com/adobe/aem/commons/assetshare/content/impl/datasources/AssetRenditionsDataSource.java [hunk 2]
```diff
@@ -76,32 +79,47 @@ protected final void doGet(SlingHttpServletRequest request, SlingHttpServletResp
...
+            }
         }
 
         dataSourceBuilder.build(request, data);
     }
 
+    private boolean acceptsAssetRenditionDispatcher(List<String> allowedAssetRenditionTypes, Set<String> excludeAssetRenditionDispatchers, AssetRenditionDispatcher assetRenditionDispatcher) {
+        if (!assetRenditionDispatcher.getTypes().isEmpty() &&
+                !allowedAssetRenditionTypes.isEmpty() &&
+                Collections.disjoint(assetRenditionDispatcher.getTypes(), allowedAssetRenditionTypes)) {
```
Comment (bstopp) [inline core/src/main/java/com/adobe/aem/commons/assetshare/content/impl/datasources/AssetRenditionsDataSource.java]: Can this be rewritten to have less `!` logic? Isn't the equivalent:

`assetRenditionDispatcher.getTypes().isEmpty() ||
                allowedAssetRenditionTypes.isEmpty() ||
                !Collections.disjoint(assetRenditionDispatcher.getTypes(), allowedAssetRenditionTypes)`


it's just hard to follow what's going on here.

Comment (davidjgonzalez) [reply core/src/main/java/com/adobe/aem/commons/assetshare/content/impl/datasources/AssetRenditionsDataSource.java]: You'd have to NOT that entire condition; which to me is even harder to grok. 
I'll add a comment explaining this condition to help mitigate confusion.

File: core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/AssetRenditionImpl.java (hunk 1 above)
Comment (davidjgonzalez) [reply core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/AssetRenditionImpl.java]: Agh - let me sort out the Url escaping better/more centrally ... the escaping is a bit of a mess. 
This *works* but to you point, it's not obvious and looks incomplete due to other places escaping happens down the chain.

File: core/src/main/java/com/adobe/aem/commons/assetshare/components/details/impl/VideoImpl.java [hunk 3]
```diff
@@ -30,27 +35,49 @@
         resourceType = { VideoImpl.RESOURCE_TYPE },
         defaultInjectionStrategy = DefaultInjectionStrategy.OPTIONAL)
 public class VideoImpl extends AbstractEmptyTextComponent implements Video {
```
[the remark this answers is not in the document]
Comment (davidjgonzalez) [reply core/src/main/java/com/adobe/aem/commons/assetshare/components/details/impl/VideoImpl.java]: True - I'll hold off on this one for now though - most of the dupe is going to go once the legacy stuff is removed. 
The dupe is more of a "coincidence" anyhow, rather than planned.

File: core/src/main/java/com/adobe/aem/commons/assetshare/components/details/impl/ImageImpl.java [hunk 4]
```diff
@@ -86,27 +108,44 @@ public void init() {
...
+                src = assetRenditions.getUrl(request, asset, parameters);
+            } else if (src == null) {
+                src = getLegacySrc();
             }
 
             if (StringUtils.isBlank(src)) {
                 src = fallbackSrc;
             }
+
+            src = StringUtils.replace(src, "%20", " ");
```
[the remark this answers is not in the document]
Comment (davidjgonzalez) [reply core/src/main/java/com/adobe/aem/commons/assetshare/components/details/impl/ImageImpl.java]: see other comment - will clean up URL escaping. agree this is confusing at best.

File: core/pom.xml [hunk 5]
```diff
@@ -37,6 +37,10 @@
     <name>Asset Share Commons - Core</name>
 
     <dependencies>
+	     <dependency>
+    		    <groupId>javax.annotation</groupId>
+		        <artifactId>javax.annotation-api</artifactId>
+        </dependency>
```
Comment (bstopp) [inline core/pom.xml]: Format your file!

File: core/src/test/java/com/adobe/aem/commons/assetshare/util/UrlUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        assertEquals("/content/dam/test%20folder/ir%c4%81%2bpu%20p%c3%b6%20%26p%c3%aep%e2%98%83.jpeg/_jcr_content/test",
+                UrlUtil.escape("/content/dam/test folder/irā+pu pö &pîp☃.jpeg/jcr:content/test"));
+    }
+
+    @Test
+    public void isEscaped() {
+        assertTrue(UrlUtil.isEscaped("/content/dam/test.png"));
+        assertTrue(UrlUtil.isEscaped("/content/dam/test%20asset.png"));
+        assertTrue(UrlUtil.isEscaped("/content/dam/test%20asset.png/_jcr_content"));
+        assertTrue(UrlUtil.isEscaped("/content/dam/test%20folder/ir%c4%81%2bpu%20p%c3%b6%20%26p%c3%aep%e2%98%83.jpeg/_jcr_content/test"));
```
Comment (bstopp) [inline core/src/test/java/com/adobe/aem/commons/assetshare/util/UrlUtilTest.java]: You need a test case for just _jcr_content without any other character escapes.
